internal-dns/src/config.rs

```diff
@@ -0,0 +1,690 @@
+    ///
+    /// This function fails only if the given sled has already been added to the
+    /// configuration.
+    pub fn host_sled(
+        &mut self,
+        sled_id: Uuid,
+        addr: Ipv6Addr,
+    ) -> anyhow::Result<Sled> {
+        match self.sleds.insert(sled_id, addr) {
+            None => Ok(Sled(sled_id)),
+            Some(existing) => Err(anyhow!(
+                "multiple definitions for sled {} (previously {}, now {})",
+                sled_id,
+                existing,
+                addr,
+            )),
+        }
+    }
+
+    /// Add a new host of type "zone" to the configuration
+    ///
+    /// Returns a [`Zone`] that can be used with [`Self::service_backend_zone()`] to
+    /// specify that this zone is a backend for some higher-level service.
+    ///
+    /// # Errors
+    ///
+    /// This function fails only if the given zone has already been added to the
+    /// configuration.
+    pub fn host_zone(
+        &mut self,
+        zone_id: Uuid,
+        addr: Ipv6Addr,
+    ) -> anyhow::Result<Zone> {
+        match self.zones.insert(zone_id, addr) {
+            None => Ok(Zone(zone_id)),
+            Some(existing) => Err(anyhow!(
+                "multiple definitions for zone {} (previously {}, now {})",
+                zone_id,
+                existing,
+                addr
+            )),
+        }
+    }
+
+    /// Specify that service `service` has a backend instance running in the
+    /// given (zone) host.
+    ///
+    /// # Errors
+    ///
+    /// This function fails only if the given host has already been added as a
+    /// backend for this service.
+    pub fn service_backend_zone(
+        &mut self,
+        service: SRV,
+        zone: &Zone,
+        port: u16,
+    ) -> anyhow::Result<()> {
+        // Although one can only get a `Zone` by adding it to a
+        // `DnsConfigBuilder`, it's possible that it was added to a different
+        // DnsBuilder.
+        ensure!(
+            self.zones.contains_key(&zone.0),
+            "zone {:?} has not been defined",
+            zone.0
+        );
+
+        let set = self
+            .service_instances_zones
+            .entry(service.clone())
+            .or_insert_with(BTreeMap::new);
+        let zone_id = zone.0;
+        match set.insert(zone_id, port) {
+            None => Ok(()),
+            Some(existing) => Err(anyhow!(
+                "service {}: zone {}: registered twice \
+                (previously port {}, now {})",
+                service.dns_name(),
+                zone_id,
+                existing,
+                port
+            )),
+        }
+    }
+
+    /// Specify that service `service` has a backend instance running directly
+    /// on the given (sled) host (generally in the sled's global zone)
+    ///
+    /// # Errors
+    ///
+    /// This function fails only if the given host has already been added as a
+    /// backend for this service.
+    pub fn service_backend_sled(
```

Comment:
I think it's only used in tests today, but it's here because at least in principle, in the spirit of RFD 248, each sled agent should be its own (SRV) service with exactly one backend:

> For something like Sled Agent, each Sled Agent would be its own service with exactly one backend, since the Sled Agents cannot be treated as interchangeable with each other.

(from RFD 248)

Today, we're not defining those SRV services at all.  This may not be a big deal.  In Nexus, this works because Nexus bypasses DNS to locate sled agents.  It uses the IP address and port stored in the database.  I'm not sure who else uses the Sled Agent client and how they look it up.

Anyway, to your question: this is here to eventually support defining an SRV service for each sled agent.

Reply:
I didn't really address the second part of your question: I believe that we do already add a AAAA record for the sled and that it's the address on the underlay network...but now I'm realizing maybe I'm wrong about which network that one is on.